### ISSUE TITLE
API functionality

### ISSUE THREAD

**Issue by louislam** [louislam/uptime-kuma#118]
### Discussed in https://github.com/louislam/uptime-kuma/discussions/96

<div type='discussions-op-text'>

<sup>Originally posted by **TheGuyDanish** July 12, 2021</sup>
I'd love to see the ability to retrieve, update, remove or add monitors via API calls. Maybe using an API key that's tied to the user. In a way that when spinning up a new docker container or doing an automated workflow like Terraform, the new device could be added to the software automatically.</div>

**Comment by Panja0** [comment]
API call for putting monitors on pause/unpause for an xx period would be awesome!
We can than include the API call in a script before doing a restart on a server for instance.

**Comment by bsord** [comment]
Commenting as I too would love to see this functionality.

**Comment by deefdragon** [comment]
Some things, like that which is visible on the status page, are already readily accessible, just undocumented. The creation and deletion of monitors/notifications are some of the things that are not. 

Thus this realistically sounds like a 2 part task.

1. Add the ability to use a key for api calls.
2. Document the API being used so that it can be easily interacted with.

**Comment by louislam** [comment]
> Some things, like that which is visible on the status page, are already readily accessible, just undocumented. The creation and deletion of monitors/notifications are some of the things that are not.
> 
> Thus this realistically sounds like a 2 part task.
> 
> 1. Add the ability to use a key for api calls.
> 2. Document the API being used so that it can be easily interacted with.

No, it should not be used in my opinion.

It is not the same thing, because these are internal APIs which I expected that it should be called from Uptime Kuma only. I could change API without any notices.

**Comment by deefdragon** [comment]
While I understand that the current calls are expected to be Uptime Kuma only, I really think it should be a single API instead of a public API and an internal API. The biggest reason is that maintaining multiple APIs for the same thing is a real pain. (I have personal experience on this from work. We had internal and external APIs for some products, and about one in 5 feature changes resulted in bugs, as changes were made to one API, but done incorrectly, or not at, all to the other).

Having a unified API has some other benefits as well:
- It Forces good documentation of the API, useful for developers and users.
- API changes would always get multiple pairs of eyes looking over them
- Obviously, each endpoint & the logic therein only has to be implemented once

I do understand the want to be able to iterate quickly, and change endpoints. This is still possible if the endpoints are versioned: alpha, beta, stable, deprecated etc. and would just require having a system around these changes. Something that should be considered anyways.

The biggest downside I can see is that the auth logic gets more complicated to account for the different ways the API is getting accessed. Given most of the work would be needed with separate APIs anyway, its not the worst thing.

**Comment by nicjansma** [comment]
This is a fantastic project, and I'm really interested in integrating Kuma into Home Assistant and other home-network things.  I'd like an API for:

* Adding / removing monitors (I have hundreds of things I want to monitor and doing it by hand would be tedious)
* Checking the overall status of how many monitors are down
* Checking the status of specific monitors

@louislam would you be up for choosing a few core APIs that would be public and supported as such?

**Comment by jesperordrup** [comment]
> Yes, I see your point.
> 
> I won't implement two sets of APIs, but as of 1.8.0, I labeled them as "internal APIs" that everyone should not use. In the future, I may re-label them as "public APIs" and documented it in the wiki once it get stable.

I would definatly chose otherwise. If you make your internal api and external api the same, you effectly put a stop to rapid development of solution. 

I would  countinue with the intenal api (for sure)
I would create a the public api as a versionised api /api/v1/methods 

Your internal api can expand change and do whatever you need in the speed you need.

The public api - of course just uses the nternal
The public api carefully ensures that it doesnt change / expose features you dont want to expose.

**Comment by jiri** [comment]
I completely agree with the point that if providing a unified API isn't something that can be done fast, it's not worth the benefit. At this point, not having an API renders Uptime Kuma borderline unusable for me, because I frequently get notification of outages whenever a scheduled service update is in place -- being able to pause an unpause monitors with an API call would alleviate this issue completely.

This issue will soon have been opened for a year and new issues requesting an API keep being opened. I hardly see a core set of functionality (like CRUD on monitors) changing in the foreseeable future and I don't see a reason for stalling just because of a private API that doesn't affect regular users.

**Comment by RedLeeder** [comment]
I agree with the overall sentiment here. Having an accessible API is extremely helpful to developers and provides a ton of out-of-the-box functionality that you otherwise wouldn't have. I would suggest you reconsider the reluctance to implement this as it is a fairly large omission from the project as it stands now (to the point where, despite the otherwise great implementation, I personally would look for alternatives).

**Comment by alichaudry** [comment]
I love Uptime-Kuma, but it sounds like it's a no-go for a lot of folks here because of its lack of API support.

Something similar to Uptime-Kuma (for my purposes, anyway) while also allowing API access is [Gatus](https://github.com/TwiN/gatus). Might be worth looking into for folks who cannot live without API access to their health + monitoring tool.

(Disclaimer: I'm using both Uptime-Kuma + Gatus as I love exploring cool new homelab projects, but I can't wait to rely solely on Uptime-Kuma once API functionality is introduced into the project.)

**Comment by anthosz** [comment]
> I love Uptime-Kuma, but it sounds like it's a no-go for a lot of folks here because of its lack of API support.
> 
> Something similar to Uptime-Kuma (for my purposes, anyway) while also allowing API access is [Gatus](https://github.com/TwiN/gatus). Might be worth looking into for folks who cannot live without API access to their health + monitoring tool.
> 
> (Disclaimer: I'm using both Uptime-Kuma + Gatus as I love exploring cool new homelab projects, but I can't wait to rely solely on Uptime-Kuma once API functionality is introduced into the project.)

Seems a read API only on Gatus so the issue is the same

**Comment by GVALFER** [comment]
i really consider this feature (API) very important. I really love thhe Kuma, but without API, it can not work for us :-(

**Comment by Computroniks** [comment]
I think that something that could be useful here is creating a list of the core APIs that would be useful and that people want. Then we could go about creating a spec using something like the OpenAPI fomat (https://swagger.io/specification/) After that, if we decide a standard way in which endpoints should be implemented, we could then divide up the endpoints between developers to distribute the workload more.

**Comment by jesperordrup** [comment]
@louislam any updates on how you think about this?

**Comment by anthosz** [comment]
> I think that something that could be useful here is creating a list of the core APIs that would be useful and that people want. Then we could go about creating a spec using something like the OpenAPI fomat (https://swagger.io/specification/) After that, if we decide a standard way in which endpoints should be implemented, we could then divide up the endpoints between developers to distribute the workload more.
> 
> Using OpenAPI has the added benefit that it can be used to generate documentation that can provide examples of how to interface with the API. There are various generators to output it in pretty much any format you like, e.g PDF, HTML, markdown.

imo, can be interesting to be able to:
- search probe (by name or url)
- add probe (with all parameters available in uptime-kuma)
- modify probe (search by name or url and modify all parameters available in uptime-kuma)
- delete probe (search by name or url)

**Comment by Computroniks** [comment]
I have created a gist to act as a sort of list for suggested endpoints to try and help them from getting lost among comments
https://gist.github.com/Computroniks/d04c2f42e430fd9f6656e40626cbdf0e

**Comment by rmawhinnie** [comment]
Uptime Kuma seems like a really neat project, however without an api to create/delete services, and pause/unapause  , it creates toil tasks for people :(.  keep up the good work,

**Comment by jiri** [comment]
@louislam This thread is getting somewhat derailed with most comments just emphasizing their desire for this feature. May I suggest constraining this issue to collaborators and writing down a definition-of-done for this task, so it can be tracked properly? 😃

**Comment by PhakornKiong** [comment]
> @louislam This thread is getting somewhat derailed with most comments just re-emphasizing the desire for this feature. May I suggest constraining this issue to collaborators and writing down a definition-of-done for this feature, so it can be tracked properly? 😃

+1

Would love to contribute to writing some of the API. Can we start a draft proposal using OpenAPI spec?

**Comment by Computroniks** [comment]
> Would love to contribute to writing some of the API. Can we start a draft proposal using OpenAPI spec?

@PhakornKiong I created a basic list of the endpoints people suggested (https://gist.github.com/Computroniks/d04c2f42e430fd9f6656e40626cbdf0e) and was kind of waiting for some feedback from @louislam before I went on to doing the spec. I will probably create a draft (i.e version 0.0.1) spec at the weekend and put it in with the gist so that I can perhaps get some feedback on the parameters of the different endpoints.

**Comment by bluepuma77** [comment]
Would love to use `uptime-kuma`, but need an API to add 100+ services with URL.

"Add foundation for API development" https://github.com/louislam/uptime-kuma/pull/1834 seems to be stuck. It only has GET, no POST.

How can we support?

**Comment by lucasheld** [comment]
I have developed an Ansible collection of modules to configure Uptime Kuma with Ansible. Currently monitors, notifications, proxies, tags, monitor tags and status pages are supported.
https://github.com/lucasheld/ansible-uptime-kuma

Under the hood, the Socket.IO API is used to communicate with Uptime Kuma.
The Ansible modules are using a Python wrapper to communicate with the API. This Python module can also be used without Ansible.
https://github.com/lucasheld/uptime-kuma-api

**Comment by lionep** [comment]
Nice work @lucasheld, the clean way to proceed IMO

**Comment by vimarshacooray** [comment]
> I have developed an Ansible collection of modules to configure Uptime Kuma with Ansible. Currently monitors, notifications, proxies, tags, monitor tags and status pages are supported.

@lucasheld Thank you!!!

this is the only reason i needed an api on kuma.

**Comment by openstep** [comment]
> > Some things, like that which is visible on the status page, are already readily accessible, just undocumented. The creation and deletion of monitors/notifications are some of the things that are not.
> > Thus this realistically sounds like a 2 part task.
> > 
> > 1. Add the ability to use a key for api calls.
> > 2. Document the API being used so that it can be easily interacted with.
> 
> No, it should not be used in my opinion.
> 
> It is not the same thing, because these are internal APIs which I expected that it should be called from Uptime Kuma only. I could change API without any notices.

Could you please @louislam add here a link to where the documentation about available APIs can be found?
I would like to know if there is fir instance an API to create an Uptime Kuma user etc.

**Comment by zoic21** [comment]
Hello,
Any new ? I'am very interest to enable/disable probe by api.

**Comment by k1m0ch1** [comment]
idk if this will help, but I desperately need the web API for uptime kuma (since my CICD using golang so I need this)

and here we go
https://github.com/k1m0ch1/kuma-uptime-webapi This is actually using @lucasheld wrapper https://github.com/lucasheld/uptime-kuma-api

thanks man @lucasheld

**Comment by MedAziz11** [comment]
I have developed a REST API wrapper for kuma since it is much needed in my work , with most of its features using the package developed by @lucasheld https://github.com/lucasheld/uptime-kuma-api .
link : https://github.com/MedAziz11/Uptime-Kuma-Web-API

**Comment by rwjack** [comment]
Now that we have maintenance mode, I think the API becomes quite a needed feature.

For example, update scripts could call the API to enable/disable maintenance mode or pause/resume individual monitors. 

The API could also be used to quickly create/remove monitors. A bash script could be made, essentially allowing the user to have a "monitor configuration file", which would be good for monitor versioning. If the file changes, the API is called and updates the monitors accordingly.

**Comment by Kovah** [comment]
Are there any plans to implement an API @louislam?

**Comment by pbdco** [comment]
> 

THANK YOU @MedAziz11 !!!!!!!!!!!! Your backend service for the API functionalities Works like a charm!

**Comment by e4rthdog** [comment]
I use Grafana and an API from uptime KUMA would give us the opportunity to keep Kuma simple and focused on what it wants to be, and at the same time it could be extensible to fit other needs. In Grafana i can represent the info as much stretched or compressed i need to and include info from other subsystems in order to give my users the status information i need them to have.

Hoping that we can have an API for reading info and some api endpoints for specific actions , as the ones i have read in this thread. For example the ability to pause/unpause a monitor due to server restart. There is no need to make a full API now and wait until Kuma reaches a stage where a full API maybe build.

**Comment by JaviSoto** [comment]
I am an iOS developer and would love to write an Uptime Kuma app using an API :)

**Comment by rwjack** [comment]
Keywords, "keep - kuma - simple"

**Comment by jjghali** [comment]
Any update on this?

**Comment by stefanux** [comment]
Am 12. Juni 2023 16:08:38 MESZ schrieb Dragonator ***@***.***>:

**Comment by CrazyHenk44** [comment]
Such an awesome project! And then no API, so I can't use it.

If you are worried about not being able to change things as fast as you want, just use an api version number in the url and not support old versions. 

Then when you update the code with an incompatible chage in the API, you bump the API number and you force people to also update the calls to the api when they download the new version and you don't have to worry.

**Comment by crstin** [comment]
A CRUD api is a must have. This is very unfortunate.

**Comment by CrazyHenk44** [comment]
I just need the uptime stats to report to our customers. It's part of the SLA, so I need this automated for all our customers. With stats per day/week, 30 days and 90 days. Without that we can't switch.

**Comment by shubhankartrivedi** [comment]
An API is a must!

**Comment by storenth** [comment]
Yeah, kuma must provide users with API to parse and modify existent monitors to integrate with CI services like gitlab.

**Comment by CommanderStorm** [comment]
@storenth @shubhankartrivedi @crstin 
Please refrain from posting `+1` / requests for updates things on issues, as this makes issue-management harder.
Issues are for discussing **what needs to be done how by whom**.
[We use 👍🏻 on issues to prioritise work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.

**Comment by CommanderStorm** [comment]
@janithcooray 

> we have thousands of sites and over 4K end points to ping


With that many endpoint the current (v1) architecture likely cound not cope.
It is yet to be benchmarked how the performance upgrades in v2 have improved upon this state.

**Comment by Besstheone** [comment]
Hello,
You can check https://github.com/MedAziz11/Uptime-Kuma-Web-API
It can be a temporary solution

**Comment by janithcooray** [comment]
> @janithcooray
> 
> > we have thousands of sites and over 4K end points to ping
> 
> With that many endpoint the current (v1) architecture likely cound not cope. It is yet to be benchmarked how the performance upgrades in v2 have improved upon this state.

yes, even the dashboard is barely usable, had use status pages. only needed the notifications and a small log disaplay so that's enough for now.
but adding and deleting points is a hassle.

**Comment by CrazyHenk44** [comment]
> @storenth @shubhankartrivedi @crstin Please refrain from posting `+1` / requests for updates things on issues, as this makes issue-management harder. Issues are for discussing **what needs to be done how by whom**. [We use 👍🏻 on issues to prioritise work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.

How does that work? This feature request has almost 3 times more 👍🏻 than any other request. If you indicate you prioritise based on them, you should actually do that, right? Within reason of course. If you don't like the feature and won't implement it, just say so, so we can all move on. There is even a 3rd party that already does it, you merge that maybe. 

Hope you can reconsider, it's just such a great product and i want it to succeed.

**Comment by georgkrause** [comment]
For me its really disgusting how people are bragging here how much infrastructure they manage, eg 

> I have hundreds of things I want to monitor

but still refuse to just contribute or fund the work required to make this happen. 

@louislam Just want to let you know that all your work is appreciated and Uptime Kuma is a insanely nice thing in my toolbox, even without API. Keep on the good work!

**Comment by louislam** [comment]
@georgkrause It's ok, I can understand that not everyone can/will understand. Since most of you guys are reasonable, so it is fine. Their opinions are not important. 

Anyway, I may decide to try to convert the current undocumented Socket.io (WebSocket) API into HTTP API by wrapping a WebSocket client on server. In this way I hope it won't take too much effort. I also hope it could be released in 2.1.0. (next next version)  
 
PR: https://github.com/louislam/uptime-kuma/pull/3854

**Comment by Aj7Ay** [comment]
Is that possible to create incident using API from shellscript ?

**Comment by CommanderStorm** [comment]
Yes, see the thirdparty api for the moment https://uptime-kuma-api.readthedocs.io/en/latest/ which is mentioned above.

**Comment by Tealk** [comment]
> API call for putting monitors on pause/unpause for an xx period would be awesome! We can than include the API call in a script before doing a restart on a server for instance.

is there anything new about this, the discussion function seems to no longer exist?

I would like to control the maintenance function via API, as I could update all my services via Ansible. This could be very easily integrated

**Comment by CommanderStorm** [comment]
Please refer to https://github.com/louislam/uptime-kuma/pull/3854 our the draft/plan to implement this.
Please refer to #4500 for the things that are needed for the current release and how helping out can help get this done faster.

> [!TIP]
> In the meantime, you can temporarily rely on an external/third party tool. (Make sure to audit their software, this is not something we have capacity for):
> - cli https://github.com/BigBoot/AutoKuma?tab=readme-ov-file#kuma-cli--
> - python https://pypi.org/project/uptime-kuma-api/
> - http https://github.com/MedAziz11/Uptime-Kuma-Web-API

**Comment by Tealk** [comment]
> * http https://github.com/MedAziz11/Uptime-Kuma-Web-API

works for me at the moment, thank you

**Comment by paul-uz** [comment]
Any update on this?

**Comment by DenuxPlays** [comment]
That is really the only thing missing in uptime kuma!
I am using it for over 2 years and it works great but I recently wanted to make some workflows automated like updating services and an API would is the one thing that is missing as I still have to go and create a maintenance for example manually.
Really hoping this is a thing sometime.

**Comment by SomeoneVeryCool** [comment]
Is this something that will be added in v2?

**Comment by jamal-0x1** [comment]
Is there a status api to check the status of Uptime Kuma itself?

I guess I can use `/dashboard`

**Comment by Tealk** [comment]
I would find an API to control the checks for automation much more interesting.

**Comment by Sammeeeeeeee** [comment]
This would be fantastic

**Comment by pr1ncey1987** [comment]
Ive created this which is a bridge server sits in-between kuma and your application - gives its a more traditional REST api.
https://github.com/pr1ncey1987/uptime-kuma-api-v2

**Comment by josefglatz** [comment]
Is there actually any way in 2026-03 to pause specific monitors or groups or monitors with a specific tag via api? I run Uptime Kuma in my homelab. Whenever the main internet fails, I want to pause affected monitors. Or do I have any other way to accomplish this?

**Comment by ArniPL** [comment]
@josefglatz If you are using third-party API like [transformable-app/Uptime-Kuma-Web-API](https://github.com/transformable-app/Uptime-Kuma-Web-API), just get all monitors, filter locally by tags and then request one by one to pause them. But that's unofficial, since Uptime Kuma doesn't have any official REST API.

**Comment by CommanderStorm** [comment]
as you can read from this thread, there are a number of thrid party solutions for this.

Nobody has sat down and wrote the glue code in uptime kuma to enable a "proper" HTTP API.
We have an internal socketio api (see wiki for more or less up to date docs), you may use it.
Write a test to know when you need to update your tooling though.

**Comment by CommanderStorm** [comment]
If someone wants to contribute engineering towards this goal, I think https://www.asyncapi.com/ might be a good first step towards a properly documented API

**Comment by rkok** [comment]
I've chosen to work around this by manipulating the sqlite db directly: https://gist.github.com/rkok/86a303f0e1f1eedfe821c90c315d42f2

**Comment by huyz** [comment]
> I've chosen to work around this by manipulating the sqlite db

Nice, why not make it a proper GitHub repo?

**Comment by renttek** [comment]
@CommanderStorm  I'm not sure if AsyncAPI is the best approach here, as it is used to model asynchronous and event driven APIs, while this is looking for a synchronous / CRUD API.

I created a draft for how this could be modeled with OpenAPI (I only included maintenance related endpoints/types yet):

<details>
<summary>OpenAPI spec</summary>

```yaml
openapi: 3.1.0
info:
  title: Uptime Kuma External API
  version: 0.0.1-draft

servers:
  - url: /api/v2

security:
  - bearerAuth: []

tags:
  - name: Maintenance

paths:
  /maintenance:
    get:
      summary: List maintenance windows
      tags:
        - Maintenance
      responses:
        '200':
          description: List of maintenance windows
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceListOutput'

    post:
      tags:
        - Maintenance
      summary: Create maintenance window
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/MaintenanceCreateInput'
      responses:
        '201':
          description: Created maintenance window
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceOutput'

  /maintenance/{maintenanceId}:
    get:
      summary: Get maintenance window
      parameters:
        - $ref: '#/components/parameters/MaintenanceId'
      tags:
        - Maintenance
      responses:
        '200':
          description: Single maintenance window
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceOutput'
        '404':
          description: No maintenance with the given ID found

    patch:
      tags:
        - Maintenance
      summary: Update maintenance window
      parameters:
        - $ref: '#/components/parameters/MaintenanceId'
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/MaintenanceCreateInput'
      responses:
        '200':
          description: Updated maintenance window
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceOutput'
        '404':
          description: No maintenance with the given ID found

    delete:
      tags:
        - Maintenance
      summary: Delete maintenance window
      parameters:
        - $ref: '#/components/parameters/MaintenanceId'
      responses:
        '204':
          description: Maintenance window deleted
        '404':
          description: No maintenance with the given ID found

  /maintenance/{maintenanceId}/pause:
    post:
      tags: [ Maintenance ]
      summary: Pause maintenance window
      parameters:
        - $ref: '#/components/parameters/MaintenanceId'
      responses:
        '200':
          description: Maintenance window paused
        '404':
          description: No maintenance with the given ID found

  /maintenance/{maintenanceId}/resume:
    post:
      tags: [ Maintenance ]
      summary: Resume maintenance window
      parameters:
        - $ref: '#/components/parameters/MaintenanceId'
      responses:
        '200':
          description: Maintenance window resumed
        '404':
          description: No maintenance with the given ID found

  /maintenance/{maintenanceId}/monitors:
    get:
      summary: Get monitors assigned to the maintenance window
      parameters:
        - $ref: '#/components/parameters/MaintenanceId'
      tags:
        - Maintenance
      responses:
        '200':
          description: List of monitors assigned to the maintenance window
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceMonitorListOutput'
        '404':
          description: No maintenance with the given ID found

    post:
      summary: Assign monitors to maintenance window
      tags:
        - Maintenance
      responses:
        '200':
          description: Monitors were assigned to maintenance window
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceMonitorListInput'
        '404':
          description: No maintenance with the given ID found

  /maintenance/{maintenanceId}/status-pages:
    get:
      summary: Get status pages assigned to maintenance window
      parameters:
        - $ref: '#/components/parameters/MaintenanceId'
      tags:
        - Maintenance
      responses:
        '200':
          description: List of status pages assigned to maintenance window
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceStatusPageListOutput'
        '404':
          description: No maintenance with the given ID found

    post:
      summary: Assign status pages to maintenance window
      tags:
        - Maintenance
      responses:
        '200':
          description: Status pages were assigned to maintenance window
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/MaintenanceStatusPageListInput'
        '404':
          description: No maintenance with the given ID found

components:
  securitySchemes:
    bearerAuth:
      type: http
      scheme: bearer
      bearerFormat: API token

  parameters:
    MaintenanceId:
      name: maintenanceId
      in: path
      required: true
      schema:
        type: integer

  schemas:
    SimpleId:
      type: object
      properties:
        id:
          type: integer
      required: integer

    ListOfSimpleIds:
      type: array
      items:
        $ref: '#/components/schemas/SimpleId'

    DateRange:
      type: array
      minItems: 2
      maxItems: 2
      items:
        - type:
            - string
            - 'null'
          format: date-time
          examples:
            - '2026-04-18T15:16'

    TimezoneOption:
      anyOf:
        - const: SAME_AS_SERVER
        - type: string
          description: IANA timezone name like Europe/Berlin
      examples:
        - Europe/Berlin
        - SAME_AS_SERVER

    TimeRange:
      type: array
      minItems: 2
      maxItems: 2
      items:
        - type: object
          properties:
            hours:
              type: integer
              minimum: 0
              maximum: 23
            minutes:
              type: integer
              minimum: 0
              maximum: 59
          required:
            - hours
            - minutes

    MaintenanceBase:
      type: object
      properties:
        title:
          type: string
        description:
          type: string
        active:
          type: boolean
          default: true
        strategy:
          type: string
          enum:
            - manual
            - single
            - cron
            - recurring-interval
            - recurring-weekday
            - recurring-day-of-month
      required:
        - title
        - strategy

    MaintenanceManual:
      allOf:
        - $ref: '#/components/schemas/MaintenanceBase'
        - type: object
          properties:
            strategy:
              const: manual
          required:
            - strategy

    MaintenanceSingle:
      allOf:
        - $ref: '#/components/schemas/MaintenanceBase'
        - type: object
          properties:
            strategy:
              const: single
            dateRange:
              $ref: '#/components/schemas/DateRange'
            timezoneOption:
              $ref: '#/components/schemas/TimezoneOption'
          required:
            - strategy
            - dateRange
            - timezoneOption

    MaintenanceCron:
      allOf:
        - $ref: '#/components/schemas/MaintenanceBase'
        - type: object
          properties:
            strategy:
              const: cron
            cron:
              type: string
              description: Cron expression
              examples:
                - '* * * * *'
                - '*/15 1-12 * * *'
            durationMinutes:
              type: integer
              minimum: 1
            dateRange:
              $ref: '#/components/schemas/DateRange'
            timezoneOption:
              $ref: '#/components/schemas/TimezoneOption'
          required:
            - strategy
            - cron
            - durationMinutes
            - dateRange
            - timezoneOption

    MaintenanceRecurringInterval:
      allOf:
        - $ref: '#/components/schemas/MaintenanceBase'
        - type: object
          properties:
            strategy:
              const: recurring-interval
          required:
            - strategy
            - dateRange
            - timezoneOption

    MaintenanceRecurringDayOfWeek:
      allOf:
        - $ref: '#/components/schemas/MaintenanceBase'
        - type: object
          properties:
            strategy:
              const: recurring-weekday
            weekday:
              type: array
              minItems: 1
              maxItems: 7
              items:
                type: integer
                minimum: 1
                maximum: 7
            timeRange:
              $ref: '#/components/schemas/TimeRange'
          required:
            - strategy
            - weekday
            - timeRange
            - dateRange
            - timezoneOption

    MaintenanceRecurringDayOfMonth:
      allOf:
        - $ref: '#/components/schemas/MaintenanceBase'
        - type: object
          properties:
            strategy:
              const: recurring-day-of-month
            daysOfMonth:
              type: array
              minItems: 1
              maxItems: 32
              items:
                oneOf:
                  - type: integer
                    minimum: 1
                    maximum: 31
                  - const: lastDay1
            timeRange:
              $ref: '#/components/schemas/TimeRange'
          required:
            - strategy
            - weekday
            - timeRange
            - dateRange
            - timezoneOption

    Maintenance:
      oneOf:
        - $ref: '#/components/schemas/MaintenanceManual'
        - $ref: '#/components/schemas/MaintenanceSingle'
        - $ref: '#/components/schemas/MaintenanceCron'
        - $ref: '#/components/schemas/MaintenanceRecurringInterval'
        - $ref: '#/components/schemas/MaintenanceRecurringDayOfWeek'
        - $ref: '#/components/schemas/MaintenanceRecurringDayOfMonth'
      discriminator:
        propertyName: strategy
        mapping:
          manual: '#/components/schemas/MaintenanceManual'
          single: '#/components/schemas/MaintenanceSingle'
          cron: '#/components/schemas/MaintenanceCron'
          recurring-interval: '#/components/schemas/MaintenanceRecurringInterval'
          recurring-weekday: '#/components/schemas/MaintenanceRecurringDayOfWeek'
          recurring-day-of-month: '#/components/schemas/MaintenanceRecurringDayOfMonth'

    MaintenanceCreateInput:
      $ref: '#/components/schemas/Maintenance'

    MaintenanceOutput:
      $ref: '#/components/schemas/Maintenance'

    MaintenanceListOutput:
      type: array
      items:
        $ref: '#/components/schemas/Maintenance'

    MaintenanceMonitorListInput:
      $ref: '#/components/schemas/ListOfSimpleIds'

    MaintenanceMonitorListOutput:
      $ref: '#/components/schemas/ListOfSimpleIds'

    MaintenanceStatusPageListInput:
      $ref: '#/components/schemas/ListOfSimpleIds'

    MaintenanceStatusPageListOutput:
      $ref: '#/components/schemas/ListOfSimpleIds'
```
</details>

If that's something that is wanted, I'd be happy to put some more work into the spec and extend it for all other needs 🙂

I'd imagine the api just as another router for express and everything else is handled inside the api layer: mapping of inout/output objects, validation and interacting with the actual models/persistence (as well as broadcasting changes via the socket api)

I'd be happy for some feedback or being told that I'm totally wrong here 😁

**Comment by CommanderStorm** [comment]
Question is how do we generate/validate this openapi and how do we keep it in sync.
Have you had a look at this PR:
- https://github.com/louislam/uptime-kuma/pull/3854/

The reason why I think that documenting our socketio api without the rest path is that adding the REST path does not seem like it is very maintainable currently, but I might be wrong

**Comment by bb-666999** [comment]
bump for REST functionality. Would be very nice with openAPI and Swagger.
